active/001-tabular-msa.md

```diff
@@ -0,0 +1,292 @@
+
+   Type of sequence stored (e.g., `DNA`)
+- `shape -> (int, int)`
+
+   Tuple storing number of rows (sequences) and columns (positions)
+- `loc -> indexer`
+
+   Allows indexing via keys (like pandas)
+- `iloc -> indexer`
+
+   Allows indexing via indices (like pandas)
+
+- `metadata -> dict`
+
+   Arbitrary metadata that applies to the alignment as a whole.
+
+- `positional_metadata -> pd.DataFrame`
+
+   Arbitrary metadata that applies to the positions of the alignment as a
+   whole.
+
+## Methods
+- `__nonzero__() -> bool`
+
+   Return `True` if MSA has one or more sequences
+- `__len__() -> int`
+
+   Return number of sequences in MSA. This is equivalent to `aln.shape[0]`
+- `__iter__() -> iterator`
+
+   Iterator of `Sequence` objects in the order specified by `keys()`
+- `__reversed__() -> iterator`
+
+   Iterator of `Sequence` objects in reverse order of `keys()`
+- `__getitem__(indexer) -> TabularMSA | Sequence`
+
+   Accepts 1 or 2 axes separated by a comma: `aln['a':'d', 50]`
+
+   The rules for axes are applied from left to right. The least significant
+   axis can only restrict the type returned, it cannot take what would have
+   been `DNA` and return a `TabularMSA`. This restriction makes it easier to
+   mentally model the indexing behavior. So the following must always hold:
+
+   ```python
+   aln[X, Y] == aln[X][:, Y] # or [Y] if the result is 1D
+   ```
+
+   #### First axis:
+   If the index is a single key or an ordinal, return a `Sequence` of type
+   `dtype`:
+
+   ```python
+   aln['a'] -> DNA
+   aln[0] -> DNA
+   ```
+
+   If the index is a slice or other complex structure of keys or ordinals,
+   return `TabularMSA`:
+
+   ```python
+   aln[2:4] -> TabularMSA
+   aln['b':'e':2] -> TabularMSA
+   aln[['a', 'c']] -> TabularMSA
+   aln[[1]] -> TabularMSA
+   aln[[True, False, False, True]] -> TabularMSA
+   ```
+
+   In the event that the keys are integers, any index composed of them will be
+   assumed to be in relation to keys and not indices
+   (like `pandas.DataFrame.ix`). It is advisable to use `iloc` and `loc` in
+   these situations.
+
+   #### Second axis:
+   The second axis will accept only column indices (e.g ordinals, slices,
+   etc.). If the first axis has already narrowed the type returned to a
+   `Sequence` object, then the second axis is simply treated as an additional
+   filter to that object:
+
+   ```python
+   aln[2, 4:7] -> DNA
+   aln[0, 0] -> DNA
+   ```
+
+   Otherise:
+
+   ```python
+   aln[2:3, 4:7] -> TabularMSA
+   ```
+
+   If the type so far is a `TabularMSA` and the second axis is an ordinal, then
+   *exactly* the type of `Sequence` will be returned. The exact `dtype` is not
+   returned because the sequence does not represent a real biological sequence
+   but rather a column in the alignment. A `Sequence` object is returned
+   because it is convenient to reuse the existing numpythonic-string API that
+   `Sequence` classes posses.
+
+   ```python
+   aln[:, 3] -> Sequence
```

Comment:
Sorry for still commenting on this while you're AFK, but I had another really good reason for this behavior:

By returning a sequence, you can start to ask questions about "who" is present at a position:

`missing = aln[:, 3].gaps()`
`who = aln.keys[missing].tolist()`

That is functionality that has never existed originally and requires zero development burden to support. Composition within our own library is easily the most powerful thing we can do. This is really the same situation as when we changed an `ignore_gaps` kwarg to `ignore` and gave it a boolean vector. Suddenly everything was more powerful, AND it required less specific knowledge and maintenance on our end.

(This also supposes that `keys` becomes a property AND a numpy vector (or Series) which seems useful)